### PR TITLE
32090 Rework logic to fetch IDPs for the login page

### LINKED ISSUE #32090: [Scalability of IDPs] Improve logic that fetches the IDPs available for login
### Description

The current logic to select the IDPs that are available to be displayed for a user during login is very complex and spread across different constructs (LoginFormsUtil, IdentityProviderBean, OrgAwareIdentityProviderBean, etc), where each construct adds another filter to be applied to the list of all IDPs that are fetched from the realm.

Ideally this should be more centralized, and use the new IDP provider to fetch the available IDPs from the DB.

### PR DESCRIPTION
**Pull request by sguilhen**
Closes #32090

### CODE REVIEW

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/idp/InfinispanIDPProvider.java [hunk 1]
```diff
@@ -143,7 +143,7 @@ public Stream<IdentityProviderModel> getAllStream(String search, Integer first,
     }
 
     @Override
-    public Stream<IdentityProviderModel> getAllStream(Map<String, String> attrs, Integer first, Integer max) {
```
Comment (pedroigor) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/idp/InfinispanIDPProvider.java:146]: Revert

File: server-spi/src/main/java/org/keycloak/models/IdentityProviderModel.java [hunk 2]
```diff
@@ -41,9 +41,12 @@ public class IdentityProviderModel implements Serializable {
     public static final String ENABLED = "enabled";
     public static final String FILTERED_BY_CLAIMS = "filteredByClaim";
     public static final String FIRST_BROKER_LOGIN_FLOW_ID = "firstBrokerLoginFlowId";
-    public static final String HIDE_ON_LOGIN = "hideOnLoginPage";
+    public static final String HIDE_ON_LOGIN = "hideOnLogin";
+    public static final String LEGACY_HIDE_ON_LOGIN_ATTR = "hideOnLoginPage";
```
Comment (pedroigor) [inline server-spi/src/main/java/org/keycloak/models/IdentityProviderModel.java:45]: Deprecated

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java [hunk 3]
```diff
@@ -160,7 +160,7 @@ public Map<String, String> parseConfig(KeycloakSession session, String config) {
                     for (AttributeType attribute : entityType.getExtensions().getEntityAttributes().getAttribute()) {
                         if (MACEDIR_ENTITY_CATEGORY.equals(attribute.getName())
                                 && attribute.getAttributeValue().contains(REFEDS_HIDE_FROM_DISCOVERY)) {
-                            samlIdentityProviderConfig.setHideOnLogin(true);
+                            samlIdentityProviderConfig.getConfig().put(IdentityProviderModel.LEGACY_HIDE_ON_LOGIN_ATTR, Boolean.TRUE.toString());
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:163]: Encapsulate in SAMLProviderConfig

File: services/src/main/java/org/keycloak/forms/login/freemarker/model/IdentityProviderBean.java [hunk 4]
```diff
@@ -97,12 +128,121 @@ private String getLoginIconClasses(IdentityProviderModel identityProvider) {
...
+     * </p>
+     * Returning an empty set essentially narrows the list of available IDPs to zero, so no IDPs will be shown for login.
+     *
+     * @param session a reference to the {@link KeycloakSession}.
+     * @param realm a reference to the realm.
+     * @param context a reference to the {@link AuthenticationFlowContext}.
+     * @return a {@link Set} containing the aliases of the IDPs that should be available for login. An empty set indicates
+     * that no IDPs should be available.
+     */
+    protected Set<String> getFederatedIdentities(KeycloakSession session, RealmModel realm, AuthenticationFlowContext context) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/forms/login/freemarker/model/IdentityProviderBean.java:182]: Actually returning th alias for linked brokers

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernameForm.java [hunk 5]
```diff
@@ -17,26 +17,28 @@
...
 public final class UsernameForm extends UsernamePasswordForm {
 
     @Override
     public void authenticate(AuthenticationFlowContext context) {
         if (context.getUser() != null) {
             // We can skip the form when user is re-authenticating. Unless current user has some IDP set, so he can re-authenticate with that IDP
-            Stream<IdentityProviderModel> identityProviders = LoginFormsUtil
-                    .filterIdentityProviders(context.getRealm().getIdentityProvidersStream(), context.getSession(), context);
-            if (identityProviders.findAny().isEmpty()) {
+            if (!this.contextUserHasFederatedIDPs(context)) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernameForm.java:41]: I don't have a better name but perhaps something like `hasLinkedBrokers`.

File: services/src/main/java/org/keycloak/forms/login/freemarker/model/IdentityProviderBean.java [hunk 6]
```diff
@@ -97,12 +128,121 @@ private String getLoginIconClasses(IdentityProviderModel identityProvider) {
...
-    public boolean isDisplayInfo() {
-        return realm.isRegistrationAllowed() || displaySocial;
+    /**
+     * Returns a predicate that can filter out IDPs associated with the current user's federated identities before those
+     * are converted into {@link IdentityProvider}s. Subclasses may use this as a way to further refine the IDPs that are
+     * to be returned.
+     *
+     * @return the custom {@link Predicate} used as a last filter before conversion into {@link IdentityProvider}
+     */
+    protected Predicate<IdentityProviderModel> federatedProviderPredicate() {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/forms/login/freemarker/model/IdentityProviderBean.java:225]: Let's see if we can use the solution we discussed about having the login filters in a single place to make it easier to identify where brokers are filtered during the login as well as help when changing those filters.

**Comment by pedroigor** [comment]
@sguilhen If you want to create a follow-up issue to remove the `BROKER_PUBLIC` when configuring brokers in the context of an organization, please do it. As discussed, it should simplify even further the implementation.

**Review by pedroigor — DISMISSED** [review]

**Comment by sguilhen** [comment]
> @sguilhen If you want to create a follow-up issue to remove the `BROKER_PUBLIC` when configuring brokers in the context of an organization, please do it. As discussed, it should simplify even further the implementation.

Done: https://github.com/keycloak/keycloak/issues/32209 marked it as important and added it to the Core IAM project.

**Review by pedroigor — APPROVED** [review]

**Review by edewit — APPROVED** [review]

**Pull request merged**